sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -848,6 +850,37 @@ def test_early_stopping_on_test_set_with_warm_start():
     gb.fit(X, y)
 
 
+def test_early_stopping_with_sample_weights(monkeypatch):
```

Comment:
This test is useful in itself but it does not check that we do not call the original `hist._raw_predict` when doing the early stopping checks.

I think the test could be extended to monkeypatch with a second mocker to check that that enabling early stopping does not induce extra calls to `hist._raw_predict` compared to fitting with a fixed number of iteration with early stopping disabled and with the default `scoring` parameter value.

Reply:
In https://github.com/scikit-learn/scikit-learn/pull/26163/commits/edafefab46a0da87522f4341d0e3a48b3f75d6e0, I added a mock to `_raw_predict` to make sure it is not called.

> compared to fitting with a fixed number of iteration with early stopping disabled and with the default scoring parameter value.

This is unclear to me. Assuming `warm_start=False`:

- With `scoring="loss"`, `_raw_predict` is not called.
- With `early_stopping` disabled, `_raw_predict` is not called.
- After this PR, `_raw_predict` is only called when `scoring` is a custom callable and early stopping is enabled. I added a new test in https://github.com/scikit-learn/scikit-learn/pull/26163/commits/b4cc72733fb4f3b6fdad9d686310f4db4de2a5a0 to assert this  behavior.